### PR TITLE
Integrated WKD in the app

### LINKED ISSUE #1201: search pubic keys of recipients on WKD
part of https://github.com/FlowCrypt/flowcrypt-android/issues/1298

trust it more than results on Attester (search first on WKD, then on Attester if WKD doesn't return public key)

1) ~port `Wkd` class from browser extension~ done
2) ~add tests for it independently of the rest of the codebase, merge the class unused~ done
3) another PR to start using this class during public key lookups - this issue

The WKD class must be **unaffected** by `disallow_attester_search_for_domains`

### PR DESCRIPTION
**Pull request by DenBond7**
This PR added support of WKD in the app

close #1201

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities

### CODE REVIEW

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/wkd/WkdClient.kt
Comment (DenBond7) [reply FlowCrypt/src/main/java/com/flowcrypt/email/api/wkd/WkdClient.kt:49]: `.usableForEncryption` - it seems that we don't have such code. How do we determine that?

Comment (DenBond7) [reply FlowCrypt/src/main/java/com/flowcrypt/email/api/wkd/WkdClient.kt]: https://github.com/FlowCrypt/flowcrypt-android/issues/480

Comment (tomholub) [reply FlowCrypt/src/main/java/com/flowcrypt/email/api/wkd/WkdClient.kt]: I see. Let's leave `usableForEncryption` for later. For now, have WkdClient return an array, and choose the first result that is not expired and not revoked. (with a comment to issue I'm about to create)

Comment (tomholub) [reply FlowCrypt/src/main/java/com/flowcrypt/email/api/wkd/WkdClient.kt]: @DenBond7 you can use `KeyRingInfo.getEncryptionSubkeys(purpose)`, where purpose is encryption. If it returns > 1 element, then the key is usable for encryption. Same for #480 later.

File: FlowCrypt/src/test/java/com/flowcrypt/email/api/email/WkdClientTest.kt
Comment (IvanPizhenko) [inline FlowCrypt/src/test/java/com/flowcrypt/email/api/email/WkdClientTest.kt:9]: why is this completely removed? what is instead of it?

**Review by IvanPizhenko — COMMENTED** [review]
Some questions here

File: FlowCrypt/src/test/java/com/flowcrypt/email/api/email/WkdClientTest.kt
Comment (DenBond7) [reply FlowCrypt/src/test/java/com/flowcrypt/email/api/email/WkdClientTest.kt:9]: After refactoring `WkdClient` requires `Context`(Android things). For now, I can't provide `Context` for JUnit tests. But maybe I'll find out how to do that and restore this test. Anyway, most of them are covered by [CreateMessageActivityWkdTest](https://github.com/FlowCrypt/flowcrypt-android/blob/7ba8e4412def7b33035975cf657b02b315700d43/FlowCrypt/src/androidTest/java/com/flowcrypt/email/ui/activity/CreateMessageActivityWkdTest.kt)

Comment (IvanPizhenko) [reply FlowCrypt/src/test/java/com/flowcrypt/email/api/email/WkdClientTest.kt:9]: Could it be left untouched and wrapped into Context in a different class?

Comment (DenBond7) [reply FlowCrypt/src/test/java/com/flowcrypt/email/api/email/WkdClientTest.kt:9]: Unfortunately, I can't leave it as-is because it is not optimized for Android(retrofit, coroutines, and so on)

Comment (DenBond7) [reply FlowCrypt/src/test/java/com/flowcrypt/email/api/email/WkdClientTest.kt:9]: @IvanPizhenko I've restored this test with a few changes. Timeout case is covered by Android Tests.

**Review by IvanPizhenko — APPROVED** [review]

**Comment by DenBond7** [comment]
@tomholub That's an important issue. I will merge it after your approve.

**Comment by tomholub** [comment]
"after your approval" or "after you approve" :-) I'll take a look

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/FlowcryptApiRepository.kt
Comment (tomholub) [inline FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/FlowcryptApiRepository.kt:152]: `apiService` is a generic name, and `getPub` is also generic - it's not clear where the public key is being fetched from.
`attesterGetPub` or `getPubFromAttester` would be better.

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/FlowcryptApiRepository.kt
Comment (tomholub) [inline FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/FlowcryptApiRepository.kt:99]: Let's call this method `pubLookup` and get rid of the `PubLookup` class.

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/util/PubLookup.kt
Comment (tomholub) [inline FlowCrypt/src/main/java/com/flowcrypt/email/api/util/PubLookup.kt:45]: As implemented, it doesn't really reflect the code structure that I was looking for when trying to mimic `PubLookup` from the browser extension. Never mind, let's simplify. You can probably put this code in `FlowCryptApiRepository` and get rid of this class.

**Review by tomholub — COMMENTED** [review]
Suggestions below to make distinction between public key sources clearer.

**Comment by tomholub** [comment]
```
class AccountKeysInfoViewModel(application: Application) : AccountViewModel(application) {
```

This should maybe be called `AccountPublicKeyServersViewModel`. I don't have a good name for it.

**Comment by tomholub** [comment]
I have a general question. This `getPub` method (which we should call `pubLookup` instead to mimic the browser) has three usages. But none of the three usages look like they are related to fetching public keys in the message compose view.

How does the compose view fetch messages? Does it use the ContactsViewModel?

image

**Comment by tomholub** [comment]
Ok I see it now, it's the `ContactsViewModel`. I'll add a comment for clarity.

**Review by tomholub — COMMENTED** [review]
Looks good, with some minor adjustments for increased clarity.

**Comment by DenBond7** [comment]
@tomholub I've added the requested changes. Please look at them.

**Review by tomholub — APPROVED** [review]

**Pull request merged**
